File: Java/ScrambleString.java
import java.util.HashMap;
import java.util.Map;

public class ScrambleString {
    private final Map<String, Boolean> memo = new HashMap<>();

    public boolean isScramble(String s1, String s2) {
        return solve(s1, s2);
    }

    private boolean solve(String a, String b) {
        if (a.equals(b)) return true;
        String key = a + "|" + b;
        Boolean cached = memo.get(key);
        if (cached != null) return cached;

        int[] cnt = new int[26];
        for (int i = 0; i < a.length(); i++) {
            cnt[a.charAt(i) - 'a']++;
            cnt[b.charAt(i) - 'a']--;
        }
        for (int c : cnt) {
            if (c != 0) {
                memo.put(key, false);
                return false;
            }
        }
        int n = a.length();
        for (int i = 1; i < n; i++) {
            if (solve(a.substring(0, i), b.substring(0, i)) && solve(a.substring(i), b.substring(i))) {
                memo.put(key, true);
                return true;
            }
            if (solve(a.substring(0, i), b.substring(n - i)) && solve(a.substring(i), b.substring(0, n - i))) {
                memo.put(key, true);
                return true;
            }
        }
        memo.put(key, false);
        return false;
    }

    public static void main(String[] args) {
        ScrambleString s = new ScrambleString();
        System.out.println(s.isScramble("great", "rgeat")); // true
        System.out.println(s.isScramble("abcde", "caebd")); // false
        System.out.println(s.isScramble("a", "a"));         // true
        System.out.println(s.isScramble("abc", "bca"));     // true
    }
}

File: Rust/src/bin/scramble_string.rs
use std::collections::HashMap;

fn solve<'a>(a: &'a str, b: &'a str, memo: &mut HashMap<(&'a str, &'a str), bool>) -> bool {
    if a == b {
        return true;
    }
    if let Some(&v) = memo.get(&(a, b)) {
        return v;
    }
    // Anagram check.
    let mut cnt = [0i32; 26];
    for (x, y) in a.bytes().zip(b.bytes()) {
        cnt[(x - b'a') as usize] += 1;
        cnt[(y - b'a') as usize] -= 1;
    }
    if cnt.iter().any(|&c| c != 0) {
        memo.insert((a, b), false);
        return false;
    }
    let n = a.len();
    for i in 1..n {
        if (solve(&a[..i], &b[..i], memo) && solve(&a[i..], &b[i..], memo))
            || (solve(&a[..i], &b[n - i..], memo) && solve(&a[i..], &b[..n - i], memo))
        {
            memo.insert((a, b), true);
            return true;
        }
    }
    memo.insert((a, b), false);
    false
}

fn is_scramble(s1: String, s2: String) -> bool {
    let mut memo: HashMap<(&str, &str), bool> = HashMap::new();
    solve(&s1, &s2, &mut memo)
}

fn main() {
    println!("{}", is_scramble("great".into(), "rgeat".into())); // true
    println!("{}", is_scramble("abcde".into(), "caebd".into())); // false
    println!("{}", is_scramble("a".into(), "a".into()));         // true
    println!("{}", is_scramble("abc".into(), "bca".into()));     // true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert!(is_scramble("great".into(), "rgeat".into()));
        assert!(!is_scramble("abcde".into(), "caebd".into()));
        assert!(is_scramble("a".into(), "a".into()));
        assert!(is_scramble("abc".into(), "bca".into()));
    }
}
